### DISCUSSION TITLE
Possible commandline parsing bug for commands

### DISCUSSION THREAD

**Discussion by DamianReeves** [com-lihaoyi/mill#3430] [Q&A] [answered]
If I define the following command:

```scala
    def elmMake(
        debug:mainargs.Flag, 
        optimize:mainargs.Flag,   
        @mainargs.arg(doc = "Specify the name of the resulting JS file.")                     
        output:Option[String],
        @mainargs.arg(doc="You can say --report=json to get error messages as JSON.")
        report:Option[String],
        @mainargs.arg(doc = "Generate a JSON file of documentation for a package.")
        docs:Option[String],      
        elmFiles:mainargs.Leftover[String]  
    ) = T.command {
        T.log.info("Running elm make command:")
        T.log.info(s"Debug: $debug")
        T.log.info(s"Optimize: $optimize")
        T.log.info(s"Output: $output")
        T.log.info(s"Report: $report")
        T.log.info(s"Docs: $docs")
        T.log.info(s"Elm Files: ${elmFiles.value.mkString(",")}")
        
    }
```

I expect that the left overs is how I specify argument processing, and that things like: output, report, and docs are command line options. However when I call this command mill seems to be treating all my strings as positional:

```
❯ ./mill elmMake  alpha beta delta gamma
[1/1] elmMake 
Running elm make command:
Debug: Flag(false)
Optimize: Flag(false)
Output: Some(alpha)
Report: Some(beta)
Docs: Some(delta)
Elm Files: gamma
```

Am I doing something wrong here?

**Comment by lihaoyi** [comment]
This is because Mill allows positional command line arguments https://github.com/com-lihaoyi/mill/blob/e5eb6487ad781654a2221168692c85f8b877edfd/main/resolve/src/mill/resolve/Resolve.scala#L157

This is not the MainArgs default, but was grandfathered in for backwards compatibility since Mill's behavior predates mainargs. We should flip it to be consistent with mainargs and add a flag for backwards compat for anyone who needs it

**Reply by lihaoyi** [reply, marked as answer]
Should be fixed in https://github.com/com-lihaoyi/mill/pull/3431

**Reply by DamianReeves** [reply]
Thanks, looking forward to the fix. I think it will be easier to grok commands when they are fully inline with mainargs
